source/sigv4.c

```diff
@@ -436,6 +561,402 @@ static SigV4Status_t parseDate( const char * pDate,
+        encodeURI( pURI, uriLen, pBufLoc, &encodedLen, false, true );
+
+        remainingLen -= encodedLen;
+
+        if( !encodeOnce )
+        {
+            encodeURI( pBufLoc, encodedLen, pBufLoc + encodedLen, &remainingLen, false, true );
+            memmove( canonicalRequest->pBufCur + encodedLen, canonicalRequest->pBufCur, remainingLen );
+        }
+
+        canonicalRequest->pBufCur += remainingLen;
+        *( canonicalRequest->pBufCur++ ) = '\n';
+
+        canonicalRequest->bufRemaining -= remainingLen + 1;
+    }
+
+/*-----------------------------------------------------------*/
+
+    static void generateCanonicalQuery( const char * pQuery,
+                                        size_t queryLen,
+                                        canonicalContext_t * canonicalRequest )
+    {
+        size_t index, remainingLen, i = 0U;
+        char * pBufLoc, * tokenQueries, * tokenParams = NULL;
+
+        assert( pQuery != NULL );
+        assert( queryLen > 0U );
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        remainingLen = canonicalRequest->bufRemaining;
+        pBufLoc = ( char * ) canonicalRequest->pBufCur;
+
+        tokenQueries = strtok( ( char * ) pQuery, "&" );
+
+        while( tokenQueries != NULL )
+        {
+            canonicalRequest->pQueryLoc[ index ] = &tokenQueries[ 0 ];
+            tokenQueries = strtok( NULL, "&" );
+
+            index++;
+        }
+
+        qsort( canonicalRequest->pQueryLoc, index, sizeof( char * ), cmpKeyValue );
+
+        for( i = 0U; i < index; i++ )
+        {
+            tokenParams = strtok( canonicalRequest->pQueryLoc[ i ], "=" );
+
+            if( tokenParams != NULL )
+            {
+                encodeURI( tokenParams, strlen( tokenParams ), pBufLoc, &remainingLen, true, false );
+                pBufLoc += remainingLen;
+                *pBufLoc = '='; /* Overwrite null character. */
+
+                canonicalRequest->bufRemaining -= remainingLen;
+                remainingLen = canonicalRequest->bufRemaining;
+            }
+
+            tokenParams = strtok( NULL, "=" );
+
+            if( tokenParams != NULL )
+            {
+                encodeURI( tokenParams, strlen( tokenParams ), pBufLoc, &remainingLen, true, false );
+                pBufLoc += remainingLen;
+
+                canonicalRequest->bufRemaining -= remainingLen;
+                remainingLen = canonicalRequest->bufRemaining;
+            }
+
+            if( index != i + 1 )
+            {
+                *pBufLoc++ = '&';
+                *pBufLoc++ = '\0';
+                *pBufLoc++ = '\n';
+                canonicalRequest->bufRemaining -= 3;
+            }
+        }
+
+        canonicalRequest->pBufCur = pBufLoc;
+    }
+
+#endif /* #if ( SIGV4_USE_CANONICAL_SUPPORT == 1 ) */
+
+/*-----------------------------------------------------------*/
+
+static SigV4Status_t verifySigV4Parameters( const SigV4Parameters_t * pParams )
```

Comment:
I have 2 comments here, firstly a lot of error strings like this can end up taking a large amount of memory on smaller systems. Did we consider if it is possible to implement this in such a way that we can minimize the memory footprint on smaller devices?

Second question, it is always useful when you return an error code which represents a the same granularity of knowledge about the problem that you know, that way it is possible for callers to act accordingly when they get an error . In this case we return the same error code for any parameter which is incorrect.  If we decide to just return SigV4InvalidParameter then I would get rid of all the duplicated assignments and invert the logic to set the return status to SigV4InvalidParameter at init and just set it to success in a last else?

Reply:
One further question, this is a static function defined in this file but never called?

Reply:
That is an interesting point, and something we hadn't given too much thought. Although the logging approach used was modelled a bit after coreHTTP ([`HTTPClient_ReadHeader`](https://github.com/FreeRTOS/coreHTTP/blob/4c896b98d7d9bfe2868e00982f2eaf64ec44752f/source/core_http_client.c#L2496-L2558), for example), it does seem like there could be much room for optimization here, just given the number of null check repetitions required, as one example. 

As you alluded to, the general approach we followed here was to supply detailed error logs explaining the rather vague status values returned from the library, but it might be worth expanding the range and detail of these enums, and maybe providing some sort of generic function to log them and avoid redundancy as much as possible. 

I will discuss this in some more detail and if time permits, address any refactoring decided upon independently. Let me know if you disagree with this approach or have any further suggestions on this. Thanks.